picard/webservice/__init__.py

```diff
@@ -338,9 +338,39 @@ def _send_request(self, request, access_token=None):
         request.access_token = access_token
         send = self._request_methods[request.method]
         data = request.data
+        if False and hasattr(self.manager, 'setTransferTimeout'):  # Available since Qt 5.15
+            native_timeout = True
+            request.setTransferTimeout(config.setting['network_transfer_timeout'])
+        else:
+            native_timeout = False
         reply = send(request, data.encode('utf-8')) if data is not None else send(request)
+        if not native_timeout:
+            self._start_transfer_timeout(reply)
         self._active_requests[reply] = request
 
+    def _start_transfer_timeout(self, reply):
+        if not config.setting['network_transfer_timeout']:
+            return
+        # Fallback implementation of a transfer timeout for Qt < 5.15.
+        # Aborts a request if no data gets transferred for TRANSFER_TIMEOUT milliseconds.
+        timer = QtCore.QTimer(self)
+        timer.setSingleShot(True)
+        timer.setTimerType(QtCore.Qt.PreciseTimer)
+        timer.timeout.connect(partial(self._timeout_request, reply))
+        reply.finished.connect(timer.stop)
+        reply.uploadProgress.connect(partial(self._reset_transfer_timeout, timer))
+        reply.downloadProgress.connect(partial(self._reset_transfer_timeout, timer))
+        timer.start(config.setting['network_transfer_timeout'])
+
+    @staticmethod
+    def _reset_transfer_timeout(timer, bytesTransferred, bytesTotal):
+        timer.start(config.setting['network_transfer_timeout'])
```

Comment:
We could store `config.setting['network_transfer_timeout']` as a property of `WsRequest` instead. It will be set for the starting request and then only change for new requests but I don't think it is much an issue.

```diff
diff --git a/picard/webservice/__init__.py b/picard/webservice/__init__.py
index 2f8d56ba..a1784904 100644
--- a/picard/webservice/__init__.py
+++ b/picard/webservice/__init__.py
@@ -150,6 +150,7 @@ class WSRequest(QNetworkRequest):
         self.priority = priority
         self.important = important
         self._high_prio_no_cache = False
+        self.timeout = config.setting['network_transfer_timeout']
 
         self.access_token = None
         self._init_headers()
@@ -340,16 +341,17 @@ class WebService(QtCore.QObject):
         data = request.data
         if False and hasattr(self.manager, 'setTransferTimeout'):  # Available since Qt 5.15
             native_timeout = True
-            request.setTransferTimeout(config.setting['network_transfer_timeout'])
+            request.setTransferTimeout(request.timeout)
         else:
             native_timeout = False
         reply = send(request, data.encode('utf-8')) if data is not None else send(request)
+        self._active_requests[reply] = request
         if not native_timeout:
             self._start_transfer_timeout(reply)
-        self._active_requests[reply] = request
 
     def _start_transfer_timeout(self, reply):
-        if not config.setting['network_transfer_timeout']:
+        timeout = self._active_requests[reply].timeout
+        if not timeout:
             return
         # Fallback implementation of a transfer timeout for Qt < 5.15.
         # Aborts a request if no data gets transferred for TRANSFER_TIMEOUT milliseconds.
@@ -358,13 +360,14 @@ class WebService(QtCore.QObject):
         timer.setTimerType(QtCore.Qt.PreciseTimer)
         timer.timeout.connect(partial(self._timeout_request, reply))
         reply.finished.connect(timer.stop)
-        reply.uploadProgress.connect(partial(self._reset_transfer_timeout, timer))
-        reply.downloadProgress.connect(partial(self._reset_transfer_timeout, timer))
-        timer.start(config.setting['network_transfer_timeout'])
+        reset_func = partial(self._reset_transfer_timeout, timer, timeout)
+        reply.uploadProgress.connect(reset_func)
+        reply.downloadProgress.connect(reset_func)
+        timer.start(timeout)
 
     @staticmethod
-    def _reset_transfer_timeout(timer, bytesTransferred, bytesTotal):
-        timer.start(config.setting['network_transfer_timeout'])
+    def _reset_transfer_timeout(timer, timeout, bytesTransferred, bytesTotal):
+        timer.start(timeout)
 
     @staticmethod
     def _timeout_request(reply):
```

Reply:
I first thought we might have multiple WebService instances running, but actially we should have only one, so I added a set_transfer_timeout method and update it on config change. We do the same for proxy settings. It should be fine, because even if a plugin should create other short lived instances it would work.

Mid-term we should think about having an even based setting system, so objects can subscribe to config changes. Would avoid dependencies on option UI on other business logic objects.